### PR TITLE
Enable auto-join-on-connect

### PR DESCRIPTION
**Pull request by Gargaj**
Fixes https://github.com/gitterHQ/irc-bridge/issues/62

### CODE REVIEW

File: lib/gitter-adapter.js [hunk 1]
```diff
@@ -306,6 +307,19 @@ Adapter.prototype.listUsers = function(channel, key) {
...
 };
 
+Adapter.prototype.autoJoin = function() {
+  this.gitterClient.currentUser()
+  .then(function(user) {
+    return user.rooms();
+  })
+  .then((function(rooms) {
+    rooms.forEach((function(room) {
+      if (room.oneToOne) return;
```
Comment (MadLittleMods) [inline lib/gitter-adapter.js]: Why not join the one-to-one rooms?

File: lib/gitter-adapter.js [hunk 2]
```diff
@@ -306,6 +307,19 @@ Adapter.prototype.listUsers = function(channel, key) {
   });
 };
 
+Adapter.prototype.autoJoin = function() {
+  this.gitterClient.currentUser()
+  .then(function(user) {
+    return user.rooms();
+  })
+  .then((function(rooms) {
```
Comment (MadLittleMods) [inline lib/gitter-adapter.js]: The extra parentheses are not needed and contradict the style already in place in this project.

File: lib/gitter-adapter.js (hunk 1 above)
Comment (Gargaj) [inline lib/gitter-adapter.js]: This might be my incomplete understanding of Gitter but aren't one-to-one rooms basically private messages between users?

Comment (MadLittleMods) [inline lib/gitter-adapter.js]: They are private messages between users and I was assuming you wouldn't see any new messages if you weren't in the IRC room but it seems to work just fine so your decision seems good to me now.

**Comment by MadLittleMods** [comment]
I am in 228 rooms(this includes one-to-ones) and running this branch only joined 33 rooms. On the server, I see the following rate limit error.

```
Error listing users for  MadLittleMods/some-room [Error: 429: {"error":"Too Many Requests"}]
```

Each join room call is multiple API requests(POST join + GET user) and the API limit 100/minute.

There was [some discussion previously about an API endpoint to join multiple rooms at once](https://github.com/dcsan/gitterbot/issues/105#issuecomment-129736477) but it wasn't needed in their case and we didn't implement. And we probably don't need that either.

Because we know that we are already a member of these rooms, we don't need to make a API request to join them. So we just need the parts of the `joinRoom` function that sends out the IRC commands to notify the client we are in the room and the call to `subscribeToRoom`.

---

The way I tackled this was to split  the logic from `joinRoom` into `joinRoomFromChannel` which takes a channel name and `joinRoom` which takes a `node-gitter` room object.

I can get about 43 this way but this still calls the user endpoint when we use `room.users()`. Skipping the `room.users()` call, I can join all rooms but then there aren't any users showing in the room :confused: 

The user issue also relates to https://github.com/gitterHQ/irc-bridge/pull/55

If you want to see what I ended up with, I put it in a gist, https://gist.github.com/MadLittleMods/e0d4c8bb55085fb717783552fea85b1b

I'll talk with the team to try to come up with a good way to tackle this.

**Comment by Gargaj** [comment]
One option of course is to just spread out the joins / user list fetches over time, depending on the API limit.

**Comment by MadLittleMods** [comment]
The room subscriptions will put a strain on resources on our backend and we don't have the bandwidth at the moment to refactor some things to make this viable.

@Gargaj Go with my gist and add some changes to limit it to the 10 most recent rooms you were in. You can sort by the `lastAccessTime`.

File: lib/gitter-adapter.js [hunk 3]
```diff
@@ -306,6 +315,36 @@ Adapter.prototype.listUsers = function(channel, key) {
...
+  })
+  .then(function(rooms) {
+    return rooms.map(function(room) {
+      return new GitterRoom(room, gc.client, gc.faye);
+    }).sort(function(a,b){ // sort descending (latest touched first)
+      return Date.parse(a.lastAccessTime) < Date.parse(b.lastAccessTime) ? 1 : -1;
+    });
+  })
+  .then(function(rooms) {
+    var limit = 10;
```
Comment (MadLittleMods) [inline lib/gitter-adapter.js]: Put this at the top as `var AUTOJOIN_LIMIT = 10;`

File: lib/gitter-adapter.js [hunk 4]
```diff
@@ -306,6 +315,36 @@ Adapter.prototype.listUsers = function(channel, key) {
...
+    rooms.every(function(room) {
```
Comment (MadLittleMods) [inline lib/gitter-adapter.js]: Use `roomse.slice(0, AUTOJOIN_LIMIT)` and get rid of the tricky decrementing inside the loop.

File: lib/gitter-adapter.js [hunk 5]
```diff
@@ -306,6 +316,36 @@ Adapter.prototype.listUsers = function(channel, key) {
...
+  var gc = this.gitterClient;
+
+  gc.currentUser()
+  .then(function(user) {
+    return user.rooms();
+  })
+  .then(function(rooms) {
+    return rooms.map(function(room) {
+      return new GitterRoom(room, gc.client, gc.faye);
+    }).sort(function(a,b){ // sort descending (latest touched first)
```
Comment (MadLittleMods) [inline lib/gitter-adapter.js]:
```
 // sort descending (latest touched first)
.sort(function(a, b) {
```

File: lib/gitter-adapter.js [hunk 6]
```diff
@@ -306,6 +316,36 @@ Adapter.prototype.listUsers = function(channel, key) {
...
+  })
+  .then(function(rooms) {
+    return rooms.map(function(room) {
+      return new GitterRoom(room, gc.client, gc.faye);
+    })
+    .filter(function(room) {
+      return !room.oneToOne;
+    })
+    .sort(function(a, b) { // sort descending (latest touched first)
+      return Date.parse(a.lastAccessTime) < Date.parse(b.lastAccessTime) ? 1 : -1;
```
Comment (MadLittleMods) [inline lib/gitter-adapter.js]: This does have the downside of not working with dates that are equal but this case probably won't come up and isn't critical.

Comment (Gargaj) [inline lib/gitter-adapter.js]: Well we could add a secondary sorting criteria, like alphabetical or something - but I doubt it matters much, we're talking about millisecond level so it's unlikely at best and even if it happens it's hardly an issue.

**Comment by MadLittleMods** [comment]
@Gargaj Thank you for the contribution :kissing_heart:

This has been merged into `develop` and will try to get everything into beta and a prod release soon after.

**Pull request closed without merging**